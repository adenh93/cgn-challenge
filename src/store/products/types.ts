import { ProductsResponse } from "../../types";

export const LOAD_PRODUCTS = "LOAD_PRODUCTS";
export const LOAD_PRODUCTS_SUCCESS = "LOAD_PRODUCT_SUCCESS";

export interface ProductState {
  productsResponse: ProductsResponse;
}

export interface LoadProductsAction {
  type: typeof LOAD_PRODUCTS;
  page?: number;
}

export interface LoadProductsSuccessAction {
  type: typeof LOAD_PRODUCTS_SUCCESS;
  response: ProductsResponse;
}

export type ProductActionTypes = LoadProductsAction | LoadProductsSuccessAction;
