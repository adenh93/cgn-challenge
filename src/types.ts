export interface ProductsResponse {
  next: string;
  previous: string;
  results: Array<Product>;
}

export interface Product {
  title: string;
  description: string;
  price: number;
  image: string;
}
