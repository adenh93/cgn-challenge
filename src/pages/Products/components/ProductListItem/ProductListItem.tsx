import React, { SFC, useState } from "react";
import {
  Card,
  CardContent,
  Typography,
  CardActions,
  Button,
  CircularProgress,
  Fade
} from "@material-ui/core";
import useStyles from "./styles";

export interface ProductItemProps {
  title: string;
  description: string;
  price: number;
  image: string;
}

const ProductListItem: SFC<ProductItemProps> = ({
  title,
  description,
  price,
  image
}) => {
  const { media, header, content, footer, progress } = useStyles();
  const [loaded, setLoaded] = useState<boolean>(false);

  const imageLoaded = () => {
    setLoaded(true);
  };

  return (
    <>
      {!loaded && (
        <CircularProgress thickness={2} className={progress} size={100} />
      )}
      <Fade in={loaded}>
        <Card elevation={3}>
          <img
            className={media}
            src={image}
            alt="product thumbnail"
            onLoad={imageLoaded}
          />
          <CardContent>
            <Typography className={header} variant="h6">
              {title}
            </Typography>
            <Typography className={content} variant="body1">
              {description}
            </Typography>
          </CardContent>
          <CardActions className={footer}>
            <Button size="medium" variant="contained" color="secondary">
              View Details
            </Button>
            <Button size="medium" variant="contained" color="primary">
              {price ? `Buy for $${price.toFixed(2)}` : "Free!"}
            </Button>
          </CardActions>
        </Card>
      </Fade>
    </>
  );
};

export default ProductListItem;
