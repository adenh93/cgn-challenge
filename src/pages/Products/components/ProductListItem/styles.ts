import { createStyles, Theme, makeStyles } from "@material-ui/core";
import { grey } from "@material-ui/core/colors";

const useStyles = makeStyles(({ palette, spacing }: Theme) =>
  createStyles({
    media: {
      height: 350,
      width: "100%"
    },
    header: {
      textTransform: "capitalize"
    },
    content: {
      color: palette.text.secondary,
      height: 40
    },
    footer: {
      borderTop: `1px solid ${grey[200]}`,
      padding: spacing(1)
    },
    progress: {
      marginLeft: "40%",
      marginTop: "50%"
    }
  })
);

export default useStyles;
