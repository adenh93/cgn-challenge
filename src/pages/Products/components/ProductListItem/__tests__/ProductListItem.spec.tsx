import * as React from "react";
import { create } from "react-test-renderer";
import { Product } from "../../../../../types";
import ProductListItem from "../../ProductListItem";

const product: Product = {
  title: "Title",
  description: "Description",
  price: 400,
  image: ""
};

describe("ProductListItem Component", () => {
  test("it matches the snapshot", () => {
    const component = create(<ProductListItem {...product} />);
    expect(component.toJSON()).toMatchSnapshot();
  });
});
