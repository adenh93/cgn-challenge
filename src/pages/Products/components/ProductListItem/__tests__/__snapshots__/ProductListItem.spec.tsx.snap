// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ProductListItem Component it matches the snapshot 1`] = `
Array [
  <div
    className="MuiCircularProgress-root makeStyles-progress-5 MuiCircularProgress-colorPrimary MuiCircularProgress-indeterminate"
    role="progressbar"
    style={
      Object {
        "height": 100,
        "width": 100,
      }
    }
  >
    <svg
      className="MuiCircularProgress-svg"
      viewBox="22 22 44 44"
    >
      <circle
        className="MuiCircularProgress-circle MuiCircularProgress-circleIndeterminate"
        cx={44}
        cy={44}
        fill="none"
        r={21}
        strokeWidth={2}
        style={Object {}}
      />
    </svg>
  </div>,
  <div
    className="MuiPaper-root MuiCard-root MuiPaper-elevation3 MuiPaper-rounded"
    style={
      Object {
        "opacity": 0,
        "visibility": "hidden",
      }
    }
  >
    <img
      alt="product thumbnail"
      className="makeStyles-media-1"
      onLoad={[Function]}
      src=""
    />
    <div
      className="MuiCardContent-root"
    >
      <h6
        className="MuiTypography-root makeStyles-header-2 MuiTypography-h6"
      >
        Title
      </h6>
      <p
        className="MuiTypography-root makeStyles-content-3 MuiTypography-body1"
      >
        Description
      </p>
    </div>
    <div
      className="MuiCardActions-root makeStyles-footer-4 MuiCardActions-spacing"
    >
      <button
        className="MuiButtonBase-root MuiButton-root MuiButton-contained MuiButton-containedSecondary"
        disabled={false}
        onBlur={[Function]}
        onDragLeave={[Function]}
        onFocus={[Function]}
        onKeyDown={[Function]}
        onKeyUp={[Function]}
        onMouseDown={[Function]}
        onMouseLeave={[Function]}
        onMouseUp={[Function]}
        onTouchEnd={[Function]}
        onTouchMove={[Function]}
        onTouchStart={[Function]}
        tabIndex={0}
        type="button"
      >
        <span
          className="MuiButton-label"
        >
          View Details
        </span>
      </button>
      <button
        className="MuiButtonBase-root MuiButton-root MuiButton-contained MuiButton-containedPrimary"
        disabled={false}
        onBlur={[Function]}
        onDragLeave={[Function]}
        onFocus={[Function]}
        onKeyDown={[Function]}
        onKeyUp={[Function]}
        onMouseDown={[Function]}
        onMouseLeave={[Function]}
        onMouseUp={[Function]}
        onTouchEnd={[Function]}
        onTouchMove={[Function]}
        onTouchStart={[Function]}
        tabIndex={0}
        type="button"
      >
        <span
          className="MuiButton-label"
        >
          Buy for $400.00
        </span>
      </button>
    </div>
  </div>,
]
`;
